[docs] add testing instructions for testing against ide-sidecar branches (#2307)
confluentinc/vscode: docs/CONTRIBUTING.md

Or by adding `.only` after a `describe` or `it` block in the test file:

```typescript
describe.only("Extension manifest tests", () => {
  it("should register all commands", () => {
    // test code
  });
});
```

Remember to remove `.only` after running the test and before making a PR!

#### Clicktesting the VS Code extension against the sidecar

##### Method 1:

Manually trigger the "Multi Arch Builds and Upload to GitHub Releases" promotion in the sidecar
repo. This can be found in the Semaphore UI under the "Promotions" tab. This won't actually upload
to a release, but will make the native executables available at the workflow's artifacts/ part of

> the Semaphore UI. Then you can continue on from step 3 of Method 2.

Reviewer: I would leave this section out (or keep in the `ide-sidecar` repo), since external users won't have access to the Semaphore promotions unless one of us starts the `/sem-approve` flow. (And even then, I'm not sure if they would see it or be able to interact with the manual promotions at all.)

##### Method 2:

> 1. Checkout branch you want to test against in the sidecar.

Reviewer: ```suggestion
1. Clone the [`ide-sidecar` repo](https://github.com/confluentinc/ide-sidecar) and check out the branch/commit you want to test against.
```

> 2. Run `make clean mvn-package-native-no-tests` in the sidecar to get the native executable packaged

Reviewer: For easier copying and pasting:
```suggestion
2. Within the cloned `ide-sidecar` directory, run the following `make` command to build the native executable.
```
make clean mvn-package-native-no-tests
```
```

Reviewer 2: nit: After checking out the Git repo and before running `make clean mvn-package-native-no-tests`, users should make sure they've installed the [prerequisites](https://github.com/confluentinc/ide-sidecar?tab=readme-ov-file#prerequisites-for-development).

   up in the target directory.

> 3. Copy the native executable from the sidecar's `target` directory to the `vscode/bin/` directory
>    (you can use click & drag).

Reviewer: Nit: not really necessary to mention clicking and dragging
```suggestion
3. Copy the native executable (`ide-sidecar-{version}-runner`) from the sidecar's `target` directory into the `vscode/bin/` directory.
```

Reviewer 2: Assuming that `$VS_CODE_EXTENSION_PROJECT` holds the path to the VS Code extension on the user's machine, they could run the following command in the root directory of the ide-sidecar project to copy the native executable:

`cp ./target/ide-sidecar-0.*.0-runner $VS_CODE_EXTENSION_PROJECT/bin`

> 4. In vscode's .versions/ide-sidecar.txt make sure it matches that specified version in the native
>    executable file name.

Reviewer: ```suggestion
4. Update [`.versions/ide-sidecar.txt`](https://github.com/confluentinc/vscode/blob/main/.versions/ide-sidecar.txt) to ensure it matches the specified version in the native
   executable file name. (For example, testing against `ide-sidecar-0.123.0-runner` requires `.versions/ide-sidecar.txt` to be `v0.123.0`.)
```

5. If there's a change to the REST or graphql model, grab the openapi generated file from sidecar

>    (underlying logic change doesn't require this step)

Reviewer 2: Can you mention that these files can be found in the folder `src/generated/resources`?

> 5. If there's a change to the REST or graphql model, grab the openapi generated file from sidecar
>    (underlying logic change doesn't require this step)
> 
> 6. Copy/paste that file to vscode's openapi file in
>    [`src/clients/sidecar-openapi-specs/sidecar.openapi.yaml`](../src/clients/sidecar-openapi-specs/sidecar.openapi.yaml)
>    or the schema in [`src/graphql/sidecar.graphql`](../src/graphql/sidecar.graphql).

Reviewer: What do you think about splitting this into two separate bullets/scenarios?
```suggestion
5. If there are any changes to the sidecar's OpenAPI spec, copy [`ide-sidecar/src/generated/resources/openapi.yaml`](https://github.com/confluentinc/ide-sidecar/blob/main/src/generated/resources/openapi.yaml) into [`src/clients/sidecar-openapi-specs/sidecar.openapi.yaml`](../src/clients/sidecar-openapi-specs/sidecar.openapi.yaml) and run `gulp apigen` (see more [here](https://github.com/confluentinc/vscode/blob/main/docs/CONTRIBUTING.md#updating-openapi-clients)

6. If there are any changes to the sidecar's GraphQL schema, copy [`ide-sidecar/src/generated/resources/schema.graphql`](https://github.com/confluentinc/ide-sidecar/blob/main/src/generated/resources/schema.graphql) into [`src/graphql/sidecar.graphql`](../src/graphql/sidecar.graphql) and similarly run `gulp apigen` (see more [here](https://github.com/confluentinc/vscode/blob/main/docs/CONTRIBUTING.md#updating-graphql-definitions))
```

> 7. Run `gulp build`.
> 
> 8. Confirm client code changes are expected.
> 
> 9. Check that ts is still linting properly: `gulp lint`.

Reviewer: ```suggestion
7. Run [`gulp ci`](https://github.com/confluentinc/vscode/blob/b342dbce8548b33f63e4f80e4a896f4c9c034664/Gulpfile.js#L35) to check for TypeScript and ESLint errors and confirm the extension builds successfully.
```

> 10. Start the extension dev host.

Reviewer: ```suggestion
10. Run the extension either via **Run > Start Debugging** (F5) or by packaging the .vsix and installing with `gulp clicktest`.
```

> 10. Start the extension dev host.

Reviewer 2: Thanks a lot for adding this documentation. It's extremely helpful.

In addition it, WDYT of adding a script to this repo that accepts the path to a local copy of the ide-sidecar repo as input parameter and automates these steps? It could build the sidecar, copy the native executable and the files from the `src/generated/resources` folder to the vs code extension project, re-generate the sidecar client, and generate the vsix file?

Reviewer: @Cerchie it's probably best kept in here or `ide-sidecar` (or both) for easy access to internal and external contributors since `dtx-tools` is private.

Or by adding `.only` after a `describe` or `it` block in the test file:

```typescript
describe.only("Extension manifest tests", () => {
  it("should register all commands", () => {
    // test code
  });
});
```

Remember to remove `.only` after running the test and before making a PR!

#### Clicktesting the VS Code extension against the sidecar

1. Clone the [`ide-sidecar` repo](https://github.com/confluentinc/ide-sidecar) and check out the
   branch/commit you want to test against. (Also ensure any sidecar
   [prerequisites](https://github.com/confluentinc/ide-sidecar?tab=readme-ov-file#prerequisites-for-development)
   are installed.)

> 2. Within the cloned `ide-sidecar` directory and make sure to run the following `make` command to

Reviewer: ```suggestion
2. Within the cloned `ide-sidecar` directory, run the following `make` command to
   build the native executable.
```

Or by adding `.only` after a `describe` or `it` block in the test file:

```typescript
describe.only("Extension manifest tests", () => {
  it("should register all commands", () => {
    // test code
  });
});
```

Remember to remove `.only` after running the test and before making a PR!

#### Clicktesting the VS Code extension against the sidecar

1. Clone the [`ide-sidecar` repo](https://github.com/confluentinc/ide-sidecar) and check out the
   branch/commit you want to test against. (Also ensure any sidecar
   [prerequisites](https://github.com/confluentinc/ide-sidecar?tab=readme-ov-file#prerequisites-for-development)
   are installed.)

> 2. Within the cloned `ide-sidecar` directory (make sure to install the
>    [prerequisites](https://github.com/confluentinc/ide-sidecar?tab=readme-ov-file#prerequisites-for-development)),
>    and make sure to run the following `make` command to build the native executable.

Reviewer: No need to mention the sidecar prereqs again in this item since it's linked in the item above.
```suggestion
2. Within the cloned `ide-sidecar` directory, run the following `make` command to build the native executable.
```

```bash
make clean mvn-package-native-no-tests
```

3.  If your `$VS_CODE_EXTENSION_PROJECT` variable has the path to VS Code, (check with

>     `echo $VS_CODE_EXTENSION_PROJECT`), you can run the following command:

Reviewer 2: What about something like:

3.  Please set the environment variable `$VS_CODE_EXTENSION_PROJECT` to the local directory holding [this repository](https://github.com/confluentinc/vscode), e.g.:

```bash
export VS_CODE_EXTENSION_PROJECT=~/Code/vscode
```

Next, copy the native executable to the `bin` directory in the VS Code extension project:

```bash
cp ./target/ide-sidecar-0.*.0-runner $VS_CODE_EXTENSION_PROJECT/bin
```
